### PR TITLE
Add load_local_shards to XLAShardedTensor

### PR DESCRIPTION
**Pull request by jonb377**
This change enables loading a list of local `XLAShard` on CPU into a sharded tensor's device data. To do this, the function `ShardingUtil::CreateShardedData` is added which transfers a list of local CPU shards to the devices and wraps the resulting handles in PjRtShardedData.

A few notes:
- Only non-replicated sharding is allowed to ensure the underlying shards remain consistent with their sharding (e.g. to prevent passing different tensors as shards to a replicated tensor).
- `PopulateTensorBuffer` is exposed from `tensor_util` for use in `ShardingUtil::CreateShardedData`.
- Each `XLAShard` will now also track which device it originated from.

### CODE REVIEW

File: torch_xla/csrc/tensor_util.cpp
Comment (yeounoh) [inline torch_xla/csrc/tensor_util.cpp:614]: Good, this works with replicated data -- as expected.

File: torch_xla/csrc/xla_sharding_util.cpp
Comment (yeounoh) [inline torch_xla/csrc/xla_sharding_util.cpp:491]: This is good, so we don't reinvent the wheels :) thanks.

File: torch_xla/experimental/xla_sharded_tensor.py
Comment (yeounoh) [inline torch_xla/experimental/xla_sharded_tensor.py:24]: 👍

File: torch_xla/csrc/init_python_bindings.cpp
Comment (JackCaoG) [inline torch_xla/csrc/init_python_bindings.cpp:1642]: make it more verbal, somthing along the line of `Can not load local shards into a non sharded tensor`

File: torch_xla/csrc/tensor_util.cpp
Comment (JackCaoG) [inline torch_xla/csrc/tensor_util.cpp:717]: hmm, any reason to move this function down?

Comment (jonb377) [reply torch_xla/csrc/tensor_util.cpp:717]: I had to move it out of the anonymous namespace to expose it outside of the file

File: test/spmd/test_xla_sharding.py
Comment (alanwaketan) [inline test/spmd/test_xla_sharding.py:98]: Does this test the replicated path?

Comment (jonb377) [reply test/spmd/test_xla_sharding.py:98]: The `load_local_shards` doesn't support replicated tensors - there would be a lot of overhead ensuring the input shards are actually the same.

Comment (alanwaketan) [reply test/spmd/test_xla_sharding.py:98]: I know that, and you have added a bunch of XLA_CHECK. Should we test the exception as well?

Comment (jonb377) [reply test/spmd/test_xla_sharding.py:98]: Thanks Jiewen! I didn't realize we could `assertRaises` on XLA_CHECK failures - I've added some more tests to cover the failure modes

**Review by alanwaketan — APPROVED** [review]
LGTM.

**Comment by jonb377** [comment]
I'll merge this after TPU CI completes

**Comment by yeounoh** [comment]
Let's rebase before merging 👍

**Pull request merged**
